### DISCUSSION TITLE
Npm install fails with package.json containing a private repository url with git+https

### DISCUSSION THREAD

**Discussion by akoskm** [community/community#25983] [Actions]
<p>I followed this guide:</p>
<aside class="onebox whitelistedgeneric">
  <header class="source">
      <img src="https://github.blog/wp-content/uploads/2019/01/cropped-github-favicon-512.png?fit=32%2C32" class="site-icon" width="32" height="32">
      <a href="https://github.blog/2012-09-21-easier-builds-and-deployments-using-git-over-https-and-oauth/" target="_blank" rel="nofollow noopener" title="02:55AM - 22 September 2012">The GitHub Blog – 22 Sep 12</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://github.blog/wp-content/uploads/2019/03/blog-card.png?fit=5001%2C2626" class="thumbnail" width="690" height="362"></div>

<h3><a href="https://github.blog/2012-09-21-easier-builds-and-deployments-using-git-over-https-and-oauth/" target="_blank" rel="nofollow noopener">Easier builds and deployments using Git over HTTPS and OAuth - The GitHub Blog</a></h3>

<p>We first introduced OAuth2 tokens in the GitHub API starting in v3. Unlike usernames and passwords, OAuth tokens provide additional benefits: Revocable access. Tokens can be revoked at any time from a user’s settings screen</p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>and created a personal access token for my account:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/7c5087382bd918730e98223c5faca05e1f4dea5d.png" data-download-href="/uploads/short-url/hJJJHPRh0MuUVQnj32CvXaC2TOl.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/3111627/181094414-abb1820a-c430-4731-931d-0f0e49c43914.png" alt="image" data-base62-sha1="hJJJHPRh0MuUVQnj32CvXaC2TOl" width="690" height="189" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/7/7c5087382bd918730e98223c5faca05e1f4dea5d_2_690x189.png, https://aws1.discourse-cdn.com/github/original/2X/7/7c5087382bd918730e98223c5faca05e1f4dea5d.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/7/7c5087382bd918730e98223c5faca05e1f4dea5d.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/7c5087382bd918730e98223c5faca05e1f4dea5d_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">752×207 23.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>this is what I have in my package.json in another project:</p>
<pre><code class="lang-auto">    "my-private-repo": "git+https://&lt;my-personal-access-token&gt;:x-oauth-basic@github.com/my-org/my-private-repo.git",
</code></pre>
<p>Two strange things are happening during <code>npm install</code> in my workflow:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/a/a445339609434944aee86dd042a5064e6bb65620.png" data-download-href="/uploads/short-url/nrczg5nbAWUJ1KbI3qToU00AYlW.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/3111627/181094415-8b509865-8c1f-4868-802a-e9173d6bce5c.png" alt="image" data-base62-sha1="nrczg5nbAWUJ1KbI3qToU00AYlW" width="690" height="296" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/a/a445339609434944aee86dd042a5064e6bb65620_2_690x296.png, https://aws1.discourse-cdn.com/github/optimized/2X/a/a445339609434944aee86dd042a5064e6bb65620_2_1035x444.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/a/a445339609434944aee86dd042a5064e6bb65620.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/a/a445339609434944aee86dd042a5064e6bb65620_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1186×510 148 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<ol>
<li>
<p>I receive a bunch of these <code>npm WARN tar ENOENT: no such file or directory</code> messages, I don’t know if they are related to the problem in 2.)</p>
</li>
<li>
<p>installation fails with <code>Permission denied</code></p>
</li>
</ol>
<p>Here’s my workflow file (I added <code>persis-credentials: false</code> after reading <a href="https://github.community/t/cannot-install-npm-module-from-a-github-private-repository/17211/3" class="inline-onebox">Cannot install npm module from a github private repository</a>):</p>
<pre><code class="lang-auto">name: Build
# This workflow is triggered on pushes to the repository.
on: [push]

jobs:
  build:

    services:
      mongodb:
        image: mongo:4.2
        ports:
          - 27017:27017

    runs-on: self-hosted
    container: my-org/my-org-base
    steps:
      - name: Checkout
        uses: actions/checkout@v2
        with:
          persist-credentials: false

      - name: Use Node.js
        uses: actions/setup-node@v1
        with:
          node-version: '8.17.0'

      - name: Install
        run: npm install

      - name: Bundle
        run: npm run build

      - name: Run unit tests
        env:
          MONGOLAB_URI: mongodb
        run: |
          ./version.sh
          npm run test-unit

      - name: Run client tests
        run: npm run test-client

      - name: Run integration tests
        env:
          MONGOLAB_URI: "mongodb://mongodb"
        run: npm run test-integration
</code></pre>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Let me help you troubleshoot this issue.</p>
<ol>
<li>Please run</li>
</ol>
<pre><code class="lang-auto">git config --local -l
</code></pre>
<p>and check the authorization token in <code>http.https://github.com/.extraheader=AUTHORIZATION: basic</code> .</p>
<ol start="2">
<li>Could you please run <code>npm install</code> command directly in a run script? You could store your PAT in <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets" rel="nofollow noopener">secrets</a>.</li>
</ol>
<pre><code class="lang-auto"> - name : install private github repo
   run: |
       npm install git+https://$ANOTHER_TOKEN:x-oauth-basic@github.com/my-org/my-private-repo.git
   env: 
       ANOTHER_TOKEN: ${{ secrets.YourPAT }}
</code></pre>
<ol start="3">
<li>Please try to run your job on ubuntu-latest runner , will npm install step succeed?</li>
</ol>

**Comment by akoskm** [comment]
<blockquote>
<p>and check the authorization token in <code>http.https://github.com/.extraheader=AUTHORIZATION: basic</code> .</p>
</blockquote>
<p>Because this is happening on a self-hosted runner I assume I should have this configuration on the machine that hosts the actions-runner. Is that correct?</p>

**Comment by Yanjingzhu** [comment]
<p>I suspect it is the credential stored for your git client (on your self-hosted runner machine) caused Permission denied issue when try to install your private repo.<br>
The actions/checkout@v2 will persist the auth in a git <code>extraheader</code> config in default. If you set persis-credentials: false , there will no extraheader in git config .<br>
So I want you to run <code>git config --local -l</code> to see whether there is the local git config of extraheader.  You need to change working directory to the work folder of the self-hosted runner .<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/a/a3235830b007f21dcf4b0db76f0c1287a771bdce.png" data-download-href="/uploads/short-url/nhbyhnWhmIr46ISMc6WAJvyZq3k.png?dl=1" title="runner git" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181094421-e9675374-1b8f-4262-82d2-e72b8d905e17.png" alt="runner git" data-base62-sha1="nhbyhnWhmIr46ISMc6WAJvyZq3k" width="592" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/a/a3235830b007f21dcf4b0db76f0c1287a771bdce_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">runner git</span><span class="informations">770×650 31.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>If there is no extraheader in your git config, could you please run<br>
npm install git+https://<a>:x-oauth-basic@github.com</a>/my-org/my-private-repo.git in your local machine under the work folder of your runner? What’s the result?<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/5/536e5401733cc6a49965f4f81a1223e986116afc.png" data-download-href="/uploads/short-url/bU40qXeAC8n6u30mpXyCNXozaTW.png?dl=1" title="self-runner" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181094423-c6d199b4-0292-4c35-829e-353b72282b3a.png" alt="self-runner" data-base62-sha1="bU40qXeAC8n6u30mpXyCNXozaTW" width="592" height="499" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/5/536e5401733cc6a49965f4f81a1223e986116afc_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">self-runner</span><span class="informations">836×706 29.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by akoskm** [comment]
<p>Surprisingly, this worked:</p>
<blockquote>
<p>Could you please run <code>npm install</code> command directly in a run script? You could store your PAT in <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets" rel="nofollow noopener">secrets</a>.</p>
</blockquote>
<p>Here’s the output from the build step:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/0/0a5141da5731cd083f16c903c841e2d7d1430bd6.png" data-download-href="/uploads/short-url/1tgRMNzvYuA4JkDaGJMdPPI10Oy.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/3111627/181094424-98199237-4901-4993-a682-d1f4084ef314.png" alt="image" data-base62-sha1="1tgRMNzvYuA4JkDaGJMdPPI10Oy" width="690" height="218" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/0/0a5141da5731cd083f16c903c841e2d7d1430bd6_2_690x218.png, https://aws1.discourse-cdn.com/github/optimized/2X/0/0a5141da5731cd083f16c903c841e2d7d1430bd6_2_1035x327.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/0/0a5141da5731cd083f16c903c841e2d7d1430bd6.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/0/0a5141da5731cd083f16c903c841e2d7d1430bd6_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1355×430 60.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>After this, I didn’t try <code>npm install git+https://:x-oauth-basic@github.com/my-org/my-private-repo.git</code> in my local machine under the work folder of the runner. After the above output, I expect it to be successful.</p>
<p>Here’s the output of <code>git config --local -l</code> from the <code>action-runner/_work/my-project/my-project</code> folder:</p>
<pre><code class="lang-auto">$ git config --local -l
core.repositoryformatversion=0
core.filemode=true
core.bare=false
core.logallrefupdates=true
remote.origin.url=https://github.com/akoskm/my-project
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
gc.auto=0
branch.develop.remote=origin
branch.develop.merge=refs/heads/develop
</code></pre>
